tar/tar.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Copyright 2020 The FluxCD contributors. All rights reserved.
+// Adapted from: golang.org/x/build/internal/untar
+
+// Package tar provides ways to manage tarball files.
+package tar
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type Tar struct {
+	maxUntarSize int
+}
+
+// NewTar creates a new Tar.
+func NewTar() *Tar {
+	return &Tar{
+		maxUntarSize: 0, // no limit
+	}
+}
+
+// WithMaxUntarSize sets the limit size for archives being decompressed by Untar.
+func (u *Tar) WithMaxUntarSize(max int) *Tar {
+	u.maxUntarSize = max
+	return u
+}
+
+// Untar reads the gzip-compressed tar file from r and writes it into dir.
+func (u *Tar) Untar(r io.Reader, dir string) (summary string, err error) {
```

Comment:
Can we add a `TarGzip` function that takes a directory and generates a tgz file? We have that in the tests and I think it would be nice to offer both functions in this new package.

Reply:
I created [an issue](https://github.com/fluxcd/pkg/issues/323) so this can be done as a follow-up task.